In [1]:
import os
import numpy as np
import cv2
import matplotlib.pyplot as plt
import random

dir = 'C:\\Users\\Richie Lee\\Desktop\\College\\1 San Jose State University\\9 2021 fall semester\\CMPE 188\\hw2&hw3\\photos\\train_set'
train_dir = './datasets/train_set/'

targets = ['buildings','forest','glacier','mountain','sea','street']
train_data = []
#load train set
for target in targets:
    train_path = os.path.join(train_dir,target)
    label = targets.index(target)
    for img in os.listdir(train_path):
        train_imgpath = os.path.join(train_path, img) #train_imgpath = './xxx/xxx/xxx.jpg'
        train_img = cv2.imread(train_imgpath)
        try:
            train_img = cv2.resize(train_img,(150,150)) #**very impoertant step
            train_image = np.array(train_img).flatten()
            train_data.append([train_image,label])
        except Exception as e:
            pass

len(train_data)

14338

In [2]:
random.shuffle(train_data)
train_data

[[array([143, 160, 146, ...,   1,   1,   1], dtype=uint8), 3],
 [array([ 19,  28,  41, ..., 101, 148, 176], dtype=uint8), 5],
 [array([24, 35, 33, ..., 21, 55, 61], dtype=uint8), 1],
 [array([163, 126, 130, ...,  45,  32,  30], dtype=uint8), 4],
 [array([200, 151, 105, ...,  99,  77,  49], dtype=uint8), 2],
 [array([214, 210, 209, ...,   9,  34,  38], dtype=uint8), 3],
 [array([179, 184, 182, ...,  15,  58,  55], dtype=uint8), 3],
 [array([127, 134, 143, ...,  19,  26,  35], dtype=uint8), 5],
 [array([ 85,  98, 114, ...,  96, 100, 105], dtype=uint8), 5],
 [array([24, 29, 32, ..., 19, 21, 21], dtype=uint8), 0],
 [array([217, 214, 206, ..., 120, 117, 109], dtype=uint8), 5],
 [array([16, 36, 84, ..., 13,  0, 12], dtype=uint8), 5],
 [array([239, 223, 207, ...,  97, 173, 202], dtype=uint8), 4],
 [array([ 1,  1,  1, ..., 66, 66, 66], dtype=uint8), 5],
 [array([0, 0, 0, ..., 2, 2, 2], dtype=uint8), 0],
 [array([7, 6, 8, ..., 3, 0, 2], dtype=uint8), 3],
 [array([175, 170, 155, ...,   6,  41,  

### train data has been shuffled before here 

In [3]:
features = []
labels = []

for feature, label in train_data:
    features.append(feature)
    labels.append(label)
    
labels

[3,
 5,
 1,
 4,
 2,
 3,
 3,
 5,
 5,
 0,
 5,
 5,
 4,
 5,
 0,
 3,
 1,
 0,
 5,
 2,
 4,
 4,
 4,
 5,
 2,
 1,
 1,
 1,
 1,
 3,
 0,
 0,
 2,
 0,
 4,
 3,
 2,
 5,
 3,
 0,
 2,
 3,
 4,
 2,
 1,
 1,
 5,
 0,
 5,
 2,
 2,
 3,
 3,
 0,
 0,
 3,
 1,
 5,
 2,
 1,
 0,
 1,
 3,
 5,
 2,
 0,
 3,
 2,
 4,
 1,
 4,
 1,
 2,
 2,
 4,
 0,
 4,
 3,
 0,
 5,
 2,
 4,
 0,
 0,
 3,
 4,
 2,
 0,
 2,
 4,
 2,
 1,
 0,
 0,
 1,
 4,
 4,
 3,
 0,
 5,
 4,
 3,
 2,
 0,
 3,
 5,
 3,
 2,
 5,
 3,
 3,
 1,
 4,
 2,
 5,
 5,
 3,
 1,
 4,
 2,
 2,
 4,
 0,
 0,
 1,
 0,
 5,
 5,
 5,
 0,
 2,
 5,
 4,
 5,
 4,
 2,
 5,
 4,
 5,
 3,
 5,
 4,
 0,
 0,
 5,
 4,
 0,
 1,
 4,
 3,
 4,
 4,
 3,
 3,
 3,
 4,
 4,
 4,
 3,
 5,
 1,
 5,
 5,
 3,
 5,
 0,
 4,
 5,
 2,
 0,
 2,
 2,
 4,
 4,
 1,
 2,
 2,
 4,
 1,
 2,
 1,
 0,
 0,
 1,
 0,
 0,
 4,
 3,
 3,
 0,
 3,
 3,
 2,
 5,
 4,
 3,
 0,
 5,
 2,
 2,
 2,
 5,
 3,
 3,
 0,
 2,
 2,
 3,
 5,
 0,
 3,
 4,
 2,
 4,
 1,
 2,
 0,
 5,
 2,
 5,
 0,
 2,
 0,
 4,
 0,
 1,
 4,
 2,
 2,
 3,
 4,
 4,
 5,
 1,
 2,
 2,
 2,
 4,
 3,
 4,
 2,
 3,
 0,
 5,
 1,
 4,
 3,
 4,
 4,
 1,


In [4]:
np_features = np.array(features)
np_lables = np.array(labels)

In [5]:
np_features = np_features.astype(np.int32)

In [6]:
np_features

array([[143, 160, 146, ...,   1,   1,   1],
       [ 19,  28,  41, ..., 101, 148, 176],
       [ 24,  35,  33, ...,  21,  55,  61],
       ...,
       [189, 179, 179, ..., 113, 102,  98],
       [163, 159, 158, ...,  98, 100, 100],
       [248, 170, 123, ...,  68, 102,  96]])

In [7]:
print('total data shape: ',np_features.shape)
print('total labels shape: ', np_lables.shape)

total data shape:  (14338, 67500)
total labels shape:  (14338,)


## Dementionality reduction

In [8]:
from sklearn.decomposition import PCA 
pca_3600 = PCA(n_components=3600)
np_features_pca = pca_3600.fit_transform(np_features)

In [9]:
np_features_pca.shape

(14338, 3600)

In [10]:
print('total data shape after PCA: ',np_features_pca.shape)
print('total labels shape: ', np_lables.shape)

total data shape after PCA:  (14338, 3600)
total labels shape:  (14338,)


In [11]:
print('the numbers of features have been dropped by', ((len(np_features[1])-len(np_features_pca[1]))/len(np_features[1]))*100,'%')

the numbers of features have been dropped by 94.66666666666667 %


In [29]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(np_features_pca, np_lables, test_size=0.2510810433812247)

In [30]:
X_train.shape

(10738, 3600)

In [31]:
np_features_pca.shape

(14338, 3600)

## Applying GridsearchCV & Cross-validation for svm

In [16]:
from sklearn import svm
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(svm.SVC(gamma='auto'), {
    'C': [1,10,20],
    'kernel': ['rbf','linear']
}, cv=5, return_train_score=False)
clf.fit(np_features_pca, np_lables)
clf.cv_results_

{'mean_fit_time': array([312.77512493, 249.71128678, 335.63347673, 261.17545252,
        343.58180594, 263.25737019]),
 'std_fit_time': array([ 1.92147989, 13.16734555, 10.61571275, 11.08636315,  6.81464167,
         6.43603076]),
 'mean_score_time': array([85.19174771, 36.18529143, 96.20823684, 38.33147783, 83.72925396,
        38.51222501]),
 'std_score_time': array([11.64672577,  0.7837059 , 19.03379362,  1.65711819,  1.91749653,
         1.8536868 ]),
 'param_C': masked_array(data=[1, 1, 10, 10, 20, 20],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_kernel': masked_array(data=['rbf', 'linear', 'rbf', 'linear', 'rbf', 'linear'],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'C': 1, 'kernel': 'rbf'},
  {'C': 1, 'kernel': 'linear'},
  {'C': 10, 'kernel': 'rbf'},
  {'C': 10, 'kernel': 'linear'},
  {'C': 20, 'kernel': 'rbf'},
  {'

In [18]:

clf_C = GridSearchCV(svm.SVC(gamma='auto'), {
    'C': [200,500],
    'kernel': ['rbf']
}, cv=2, return_train_score=False)
clf_C.fit(np_features_pca, np_lables)
clf_C.cv_results_

{'mean_fit_time': array([116.13888967, 118.05938911]),
 'std_fit_time': array([2.09857619, 0.47964382]),
 'mean_score_time': array([132.13365996, 119.42439389]),
 'std_score_time': array([15.36980999,  0.21228814]),
 'param_C': masked_array(data=[200, 500],
              mask=[False, False],
        fill_value='?',
             dtype=object),
 'param_kernel': masked_array(data=['rbf', 'rbf'],
              mask=[False, False],
        fill_value='?',
             dtype=object),
 'params': [{'C': 200, 'kernel': 'rbf'}, {'C': 500, 'kernel': 'rbf'}],
 'split0_test_score': array([0.20965267, 0.20965267]),
 'split1_test_score': array([0.20895522, 0.20895522]),
 'mean_test_score': array([0.20930395, 0.20930395]),
 'std_test_score': array([0.00034872, 0.00034872]),
 'rank_test_score': array([1, 1])}

In [23]:
from sklearn import svm
from sklearn.model_selection import GridSearchCV
clf_C_135 = GridSearchCV(svm.SVC(), {
    'C': [1,3,5],
    'kernel': ['rbf']
}, cv=3, return_train_score=False)
clf_C_135.fit(np_features_pca, np_lables)
clf_C_135.cv_results_

{'mean_fit_time': array([116.51370144, 136.217954  , 146.35602649]),
 'std_fit_time': array([1.65754119, 0.04390556, 4.20291625]),
 'mean_score_time': array([91.78611771, 97.07998451, 92.14324609]),
 'std_score_time': array([6.37503297, 8.53647581, 2.88359949]),
 'param_C': masked_array(data=[1, 3, 5],
              mask=[False, False, False],
        fill_value='?',
             dtype=object),
 'param_kernel': masked_array(data=['rbf', 'rbf', 'rbf'],
              mask=[False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'C': 1, 'kernel': 'rbf'},
  {'C': 3, 'kernel': 'rbf'},
  {'C': 5, 'kernel': 'rbf'}],
 'split0_test_score': array([0.65523013, 0.66861925, 0.6665272 ]),
 'split1_test_score': array([0.66059845, 0.68110483, 0.67712911]),
 'split2_test_score': array([0.65515798, 0.67419962, 0.67273488]),
 'mean_test_score': array([0.65699552, 0.67464123, 0.67213039]),
 'std_test_score': array([0.00254783, 0.00510678, 0.00434927]),
 'rank_test_score': ar

In [24]:
import pandas as pd
df = pd.DataFrame(clf_C_135.cv_results_)
df[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0       1          rbf         0.656996
1       3          rbf         0.674641
2       5          rbf         0.672130

* svm has its highest accuracy as 0.674641 when usinng rbf kernal and C = 3

## Applying GridsearchCV & Cross-validation for k-nn

In [15]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
clf_nn = GridSearchCV(KNeighborsClassifier(), {
    'n_neighbors': [1,2,3,4,5],
}, cv=5, return_train_score=False)
clf_nn.fit(np_features_pca, np_lables)
clf_nn.cv_results_

{'mean_fit_time': array([0.11903977, 0.12629099, 0.1203197 , 0.11813312, 0.11955128]),
 'std_fit_time': array([0.00412718, 0.0054918 , 0.00156083, 0.00306848, 0.00429199]),
 'mean_score_time': array([1.26897206, 1.46239386, 1.47947431, 1.59414907, 1.58798342]),
 'std_score_time': array([0.02701923, 0.02832398, 0.03985029, 0.03878916, 0.04912124]),
 'param_n_neighbors': masked_array(data=[1, 2, 3, 4, 5],
              mask=[False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'n_neighbors': 1},
  {'n_neighbors': 2},
  {'n_neighbors': 3},
  {'n_neighbors': 4},
  {'n_neighbors': 5}],
 'split0_test_score': array([0.41945607, 0.41875872, 0.43723849, 0.42956764, 0.43235704]),
 'split1_test_score': array([0.42050209, 0.41631799, 0.43200837, 0.41666667, 0.41631799]),
 'split2_test_score': array([0.41701534, 0.41213389, 0.43340307, 0.42608089, 0.42468619]),
 'split3_test_score': array([0.40844088, 0.396233  , 0.41611441, 0.41192885, 0.4066969 ]),


In [16]:

df = pd.DataFrame(clf_nn.cv_results_)
df[['param_n_neighbors','mean_test_score']]

param_n_neighbors  mean_test_score
0                 1         0.414492
1                 2         0.408284
2                 3         0.428301
3                 4         0.419444
4                 5         0.417630

* k-nn has its highest accuracy as 0.428301 when usinng n = 3

## We chose svm as classifier

* training the support vectorclassifier

In [34]:
X_test.shape

(3600, 3600)

In [37]:
from sklearn.svm import SVC
model = SVC(C=4)
model.fit(np_features_pca,np_lables)
#model.score(X_test,y_test)

SVC(C=4)

In [16]:
test_dir = './datasets/test_set/'
test_data = []
#load test set
for img in os.listdir(test_dir):
    path = os.path.join(test_dir, img)
    test_img = cv2.imread(path)
    
    try:
        test_img = cv2.resize(test_img,(150,150))
        image = np.array(test_img).flatten()
        test_data.append([image,img])
    except Exception as e:
        pass

In [38]:
len(test_data)

3600

In [18]:
test_features = []
image_ids = []

for test_feature, image_id in test_data:
    test_features.append(test_feature)
    image_ids.append(image_id)

In [39]:
test_features

[array([  4,   7,  21, ..., 109,  88,  86], dtype=uint8),
 array([222, 194, 160, ...,  75, 107, 126], dtype=uint8),
 array([250, 250, 250, ...,  59,  63,  74], dtype=uint8),
 array([104,  94,  47, ...,  67,  70,  78], dtype=uint8),
 array([241, 246, 244, ..., 124, 101, 109], dtype=uint8),
 array([153, 106,  44, ...,  30,  20,  20], dtype=uint8),
 array([230, 199, 146, ...,  80, 198, 179], dtype=uint8),
 array([95, 78, 69, ..., 42, 45, 59], dtype=uint8),
 array([ 0,  6,  4, ..., 13,  6, 13], dtype=uint8),
 array([ 30,  30,  30, ..., 100, 100, 100], dtype=uint8),
 array([ 53,  32,   4, ..., 122,  99,  73], dtype=uint8),
 array([ 95,  46,  14, ...,  56,  92, 116], dtype=uint8),
 array([162, 162, 162, ...,   7,   7,   7], dtype=uint8),
 array([ 63,  67,  72, ..., 148, 159, 167], dtype=uint8),
 array([243, 241, 240, ...,  99,  98, 107], dtype=uint8),
 array([163, 163, 163, ...,  56,  56,  56], dtype=uint8),
 array([100,  93, 120, ...,  76,  93, 120], dtype=uint8),
 array([217, 200, 197, ...

In [40]:
np_test_features = np.array(test_features)
np_test_features = np_test_features.astype(np.int32)
np_test_features.shape

(3600, 67500)

In [41]:
print('total test data shape: ',np_test_features.shape)

total test data shape:  (3600, 67500)


## Dementionality reduction for test data
## using same eigenvalue

In [22]:
np_test_features_pca = pca_3600.transform(np_test_features)
np_test_features_pca.shape

(3600, 3600)

In [43]:
svm_predict_result = model.predict(np_test_features_pca)

In [46]:
model.get_params

<bound method BaseEstimator.get_params of SVC(C=4)>

In [47]:
svm_predict_result

array([5, 0, 0, ..., 2, 1, 2])

In [48]:
import pandas as pd
df_pca_svm = pd.DataFrame()

df_pca_svm['Id'] = image_ids
df_pca_svm['Category'] = svm_predict_result
df_pca_svm.to_csv('PCA_svm_prediction_v1.9.csv')

In [49]:
import pandas as pd
csv = pd.read_csv('PCA_svm_prediction_v1.9.csv')
#csv = csv.drop(columns=['AB'])

In [50]:
csv

Unnamed: 0                          Id  Category
0              0      1000 - Copy - Copy.jpg         5
1              1                   10113.jpg         0
2              2                   10144.jpg         0
3              3                   10151.jpg         0
4              4                   10161.jpg         0
...          ...                         ...       ...
3595        3595              991 - Copy.jpg         0
3596        3596  9943 - Copy (2) - Copy.jpg         2
3597        3597  9951 - Copy (2) - Copy.jpg         2
3598        3598                     997.jpg         1
3599        3599   998 - Copy (2) - Copy.jpg         2

[3600 rows x 3 columns]

# the kaggle prediction result is 0.69960